README.md
# Janda.Parsers.DateTime

[![.NET](https://github.com/Jandini/Janda.Parsers.DateTime/actions/workflows/dotnet.yml/badge.svg)](https://github.com/Jandini/Janda.Parsers.DateTime/actions/workflows/dotnet.yml)

Provides extensions for date time parsing in byte arrays.




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.3.0](https://github.com/Jandini/Janda.Parsers.DateTime/compare/1.2.0...1.3.0) (2020-07-17)


### Features

* Add dot packages ([0cec779](https://github.com/Jandini/Janda.Parsers.DateTime/commit/0cec7793bd5bf8b56187886eb2a8769a7a9316e1))
* Add jenkins shared library to jenkins file ([f6a26bc](https://github.com/Jandini/Janda.Parsers.DateTime/commit/f6a26bc34fb113339ee80aa0abc38669d1993424))
* Bump dot packages version to 0.3.0 ([00fedde](https://github.com/Jandini/Janda.Parsers.DateTime/commit/00feddeaf32aec556cf67c9472a70bbb27ceeaf3))

## [1.2.0](https://github.com/Jandini/Janda.Parsers.DateTime/compare/1.1.0...1.2.0) (2020-06-27)

## 1.1.0 (2020-06-07)


### Features

* Add big endian hfs date parser ([0e33960](https://github.com/Jandini/Janda.Parsers.DateTime/commit/0e339609b1746fe96b2f3e8638d5a10787994777))
* Add cdate time and hfs file time parsers ([e52a681](https://github.com/Jandini/Janda.Parsers.DateTime/commit/e52a6811bdd68af2670aaec24c8486fb470718d5))
* Add cstring int cdate parser ([f8f99bf](https://github.com/Jandini/Janda.Parsers.DateTime/commit/f8f99bf8cdcb50d1e425a7082fdd3efcaa48dadc))
* Replace index based arrays operations with span and slice ([3196787](https://github.com/Jandini/Janda.Parsers.DateTime/commit/3196787d46648e1eccb9f694d1434ca0122e2c33))


src/Janda.Parsers.DateTime.Tests/ByteArrayExtenstions_MustThrow.cs
using System;
using Xunit;

namespace Janda.Parsers.Tests
{
    public class ByteArrayExtenstions_MustThrow
    {            
        [Fact]
        public void ReadAsBigEndianCDateToDateTime_WhenBufferIsEmpty_ThrowsIndexOutOfRangeException()
        {
            const int EMPTY_ARRAY_SIZE = 0;
            var buffer = new byte[EMPTY_ARRAY_SIZE];
            Assert.Throws<ArgumentOutOfRangeException>(()=>buffer.ReadAsBigEndianCDateToDateTime());
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_WhenBufferIsTooShort_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            var buffer = new byte[C_DATE_SIZE - 1];
            Assert.Throws<ArgumentOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime());
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_WhenIndexIsNegative_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            const int NEGATIVE_INDEX = -1;
            var buffer = new byte[C_DATE_SIZE];
            Assert.Throws<ArgumentOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime(NEGATIVE_INDEX));
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_WhenIndexIsOutsideOfRange_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            const int OUTSIDE_OF_RANGE_INDEX = C_DATE_SIZE + 1;
            var buffer = new byte[C_DATE_SIZE];
            Assert.Throws<ArgumentOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime(OUTSIDE_OF_RANGE_INDEX));
        }
    }
}


src/Janda.Parsers.DateTime.Tests/ByteArrayExtenstions_MustReturn.cs
using System;
using System.Buffers.Binary;
using Xunit;

namespace Janda.Parsers.Tests
{
    public class ByteArrayExtenstions_MustReturn
    {
        [Fact]
        public void ReadAsHFSDateToDateTime_AtIndexZero_ReturnsEpoch()
        {
            const int HFS_DATE_SIZE = 4;
            const int INDEX_ZERO = 0;
            var HFS_DATE_EPOCH = new DateTime(1904, 1, 1);

            var buffer = new byte[HFS_DATE_SIZE];

            Assert.Equal(buffer.ReadAsHFSDateToDateTime(INDEX_ZERO), HFS_DATE_EPOCH);
        }

        [Fact]
        public void ReadAsCDateToDateTimeReturnsEpoch()
        {
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsCDateToDateTime(), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsCDateToDateTime_AtIndexZero_ReturnsEpoch()
        {
            const int INDEX_ZERO = 0;
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsCDateToDateTime(INDEX_ZERO), C_DATE_EPOCH);
        }

        [Fact]
        public void ReadAsBigEndianCDateToDateTime_AtIndexZero_ReturnsEpoch()
        {
            const int INDEX_ZERO = 0;
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(INDEX_ZERO), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTimeReturnsEpoch()
        {
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_AtIndexZero_ReturnsExpectedDate()
        {
            const int FIRST_SECOND = 1;
            const int INDEX = 0;
            const int C_DATE_SIZE = 4;
            var FIRST_SECOND_AS_BIG_ENDIAN = new DateTime(1970, 07, 14, 4, 20, 16);

            var buffer = new byte[C_DATE_SIZE];

            BinaryPrimitives.WriteInt32LittleEndian(buffer, FIRST_SECOND);

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(INDEX), FIRST_SECOND_AS_BIG_ENDIAN);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTimeReturnsExpectedDate()
        {
            DateTime EXPECTED_DATE_TIME = new DateTime(1980, 08, 21, 19, 45, 00);
            byte[] C_DATE_BUFFER = { 0x14, 0x02, 0xE9, 0x3C };

            Assert.Equal(C_DATE_BUFFER.ReadAsBigEndianCDateToDateTime(), EXPECTED_DATE_TIME);
        }

        [Fact]
        public void ReadAsCDateToDateTimeReturnsExpectedDate()
        {
            DateTime EXPECTED_DATE_TIME = new DateTime(2002, 05, 20, 14, 03, 00);
            byte[] C_DATE_BUFFER = { 0x14, 0x02, 0xE9, 0x3C };

            Assert.Equal(C_DATE_BUFFER.ReadAsCDateToDateTime(), EXPECTED_DATE_TIME);
        }
    }
}


src/Janda.Parsers.DateTime.Tests/ByteArrayExtenstions_Must.cs
using System;
using System.Buffers.Binary;
using Xunit;

namespace Janda.Parsers.Tests
{
    public class ByteArrayExtenstions_Must
    {      
        [Fact]
        public void ReadAsHFSDateToDateTime_AtIndexZero_ReturnEpoch()
        {
            const int HFS_DATE_SIZE = 4;
            const int INDEX_ZERO = 0;
            var HFS_DATE_EPOCH = new DateTime(1904, 1, 1);
                        
            var buffer = new byte[HFS_DATE_SIZE];

            Assert.Equal(buffer.ReadAsHFSDateAsDateTime(INDEX_ZERO), HFS_DATE_EPOCH);
        }

        [Fact]
        public void ReadAsCDateToDateTime_ReturnEpoch()
        {
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);
                        
            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsCDateToDateTime(), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsCDateToDateTime_AtIndexZero_ReturnEpoch()
        {
            const int INDEX_ZERO = 0;
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsCDateToDateTime(INDEX_ZERO), C_DATE_EPOCH);
        }

        [Fact]
        public void ReadAsBigEndianCDateToDateTime_IndexZero_ReturnEpoch()
        {
            const int INDEX_ZERO = 0;
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(INDEX_ZERO), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_ReturnEpoch()
        {
            const int C_DATE_SIZE = 4;
            var C_DATE_EPOCH = new DateTime(1970, 1, 1);

            var buffer = new byte[C_DATE_SIZE];

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(), C_DATE_EPOCH);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_EmptyBuffer_ThrowsIndexOutOfRangeException()
        {
            const int EMPTY_ARRAY_SIZE = 0;
            var buffer = new byte[EMPTY_ARRAY_SIZE];
            Assert.Throws<IndexOutOfRangeException>(()=>buffer.ReadAsBigEndianCDateToDateTime());
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_TooShortBuffer_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            var buffer = new byte[C_DATE_SIZE - 1];
            Assert.Throws<IndexOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime());
        }



        [Fact]
        public void ReadAsBigEndianCDateToDateTime_NegativeIndex_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            const int NEGATIVE_INDEX = -1;
            var buffer = new byte[C_DATE_SIZE];
            Assert.Throws<IndexOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime(NEGATIVE_INDEX));
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_IndexOutsideOfRange_ThrowsIndexOutOfRangeException()
        {
            const int C_DATE_SIZE = 4;
            const int OUTSIDE_OF_RANGE_INDEX = C_DATE_SIZE + 1;
            var buffer = new byte[C_DATE_SIZE];
            Assert.Throws<IndexOutOfRangeException>(() => buffer.ReadAsBigEndianCDateToDateTime(OUTSIDE_OF_RANGE_INDEX));
        }

        [Fact]
        public void ReadAsBigEndianCDateToDateTime_AtIndexZero_ReturnExpectedDate()
        {
            const int FIRST_SECOND = 1;
            const int INDEX = 0;
            const int C_DATE_SIZE = 4;
            var FIRST_SECOND_AS_BIG_ENDIAN = new DateTime(1970, 07, 14, 4, 20, 16);

            var buffer = new byte[C_DATE_SIZE];
            
            BinaryPrimitives.WriteInt32LittleEndian(buffer, FIRST_SECOND);      

            Assert.Equal(buffer.ReadAsBigEndianCDateToDateTime(INDEX), FIRST_SECOND_AS_BIG_ENDIAN);
        }


        [Fact]
        public void ReadAsBigEndianCDateToDateTime_ReturnExpectedDate()
        {
            DateTime EXPECTED_DATE_TIME = new DateTime(1980, 08, 21, 19, 45, 00);            
            byte[] C_DATE_BUFFER = { 0x14, 0x02, 0xE9, 0x3C };            

            Assert.Equal(C_DATE_BUFFER.ReadAsBigEndianCDateToDateTime(), EXPECTED_DATE_TIME);
        }


        [Fact]
        public void ReadAsCDateToDateTime_ReturnExpectedDate()
        {
            DateTime EXPECTED_DATE_TIME = new DateTime(2002, 05, 20, 14, 03, 00);
            byte[] C_DATE_BUFFER = { 0x14, 0x02, 0xE9, 0x3C };

            Assert.Equal(C_DATE_BUFFER.ReadAsCDateToDateTime(), EXPECTED_DATE_TIME);
        }
    }
}
